cross-chain/solana/programs/tbtc/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+use anchor_lang::prelude::*;
+use anchor_spl::{
+    associated_token::AssociatedToken,
+    // mint,
+    token::{mint_to, TokenAccount, Mint as SplMint, MintTo as SplMintTo, Token as SplToken},
+};
+
+declare_id!("HksEtDgsXJV1BqcuhzbLRTmXp5gHgHJktieJCtQd3pG");
+
+#[program]
+pub mod tbtc {
+    use super::*;
+    
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+        let tbtc = &mut ctx.accounts.tbtc; 
+        tbtc.authority = ctx.accounts.authority.key();
+        tbtc.token_mint = ctx.accounts.tbtc_mint.key();
+        tbtc.token_bump = *ctx.bumps.get("tbtc_mint").unwrap();
+        tbtc.minters = 0;
+        tbtc.guardians = 0;
+        tbtc.paused = false;
+        Ok(())
+    }
+
+    pub fn add_minter(ctx: Context<AddMinter>) -> Result<()> {
+        let minter_info = &mut ctx.accounts.minter_info;
+        minter_info.minter = ctx.accounts.minter.key();
+        minter_info.bump = *ctx.bumps.get("minter_info").unwrap();
+
+        ctx.accounts.tbtc.minters += 1;
+        emit!(MinterAdded { minter: ctx.accounts.minter.key() });
+        Ok(())
+    }
+
+    pub fn remove_minter(ctx: Context<RemoveMinter>, minter: Pubkey) -> Result<()> {
+        // require!(ctx.accounts.tbtc.minters > 0, TbtcError::NoMinters);
+        ctx.accounts.tbtc.minters -= 1;
+        emit!(MinterRemoved { minter: minter });
+        Ok(())
+    }
+
+    pub fn add_guardian(ctx: Context<AddGuardian>) -> Result<()> {
+        let guardian_info = &mut ctx.accounts.guardian_info;
+        guardian_info.guardian = ctx.accounts.guardian.key();
+        guardian_info.bump = *ctx.bumps.get("guardian_info").unwrap();
+
+        ctx.accounts.tbtc.guardians += 1;
+        emit!(GuardianAdded { guardian: ctx.accounts.guardian.key() });
+        Ok(())
+    }
+
+    pub fn remove_guardian(ctx: Context<RemoveGuardian>, guardian: Pubkey) -> Result<()> {
+        // require!(ctx.accounts.tbtc.guardians > 0, TbtcError::NoGuardians);
+        ctx.accounts.tbtc.guardians -= 1;
+        emit!(GuardianRemoved { guardian: guardian });
+        Ok(())
+    }
+
+    pub fn setup_mint(_ctx: Context<SetupMint>) -> Result<()> {
+        msg!("setup mint complete\n\n\n");
```

Comment:
Write to program logs sparingly. This message doesn't cost much. But there isn't much of a point to it because you can see when an instruction is successfully invoked in the program log.